File: widget/src/main/java/com/yj/widget/recycler/RecyclerViewAdapter.java
package com.yj.widget.recycler;


import com.yj.widget.WidgetActivity;

public class RecyclerViewAdapter<T> extends BaseRecyclerViewAdapter<T> {

    private BaseRecyclerWidget<T> baseListWidget;

    public RecyclerViewAdapter(WidgetActivity activity, BaseRecyclerWidget<T> baseListWidget) {
        super(activity);
        this.baseListWidget = baseListWidget;
    }

    @Override
    protected RecyclerViewHolderWidget<T> onCreateViewHolderWidget(int viewType) {
        return baseListWidget.onCreateViewHolderWidget(viewType);
    }

    @Override
    public T getItem(int position) {
        return baseListWidget.getItem(position);
    }

    @Override
    public int getItemCount() {
        return baseListWidget.getItemCount();
    }


    @Override
    public int getItemViewType(int position) {
        return baseListWidget.getItemViewType(position);
    }

    public int getDataCount() {
        return data == null ? 0 : data.size();
    }

    public T getDataItem(int position) {
        if (position < 0 && position >= data.size()) {
            return null;
        }
        return data.get(position);

    }

}

File: widget/src/main/java/com/yj/widget/recycler/RecyclerViewAdapter.kt
package com.yj.widget.recycler

import androidx.paging.PagingDataAdapter
import androidx.recyclerview.widget.RecyclerView
import android.view.ViewGroup
import androidx.annotation.IntRange
import androidx.recyclerview.widget.DiffUtil
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers

/**
 * <pre>
 *     author: yijin
 *     date  : 2022/6/5
 *     desc  :
 * </pre>
 */
class RecyclerViewAdapter<K : Any, T : Any>(
    val recyclerWidget: RecyclerWidget<K, T>,
    diffCallback: DiffUtil.ItemCallback<T>,
    mainDispatcher: CoroutineDispatcher = Dispatchers.Main,
    workerDispatcher: CoroutineDispatcher = Dispatchers.Default
) :
    PagingDataAdapter<T, RecyclerView.ViewHolder>(diffCallback, mainDispatcher, workerDispatcher) {


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {

        val widget = onCreateViewHolderWidget(viewType)
        widget.setAdapter(this)
        recyclerWidget.loadChildWidget(
            recyclerWidget.contentView as ViewGroup,
            widget.disableAddView().get()
        )
        return RecyclerViewHolder(widget)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val viewHolder: RecyclerViewHolder<T> = holder as RecyclerViewHolder<T>
        bindData(viewHolder.widget, position)
    }

    override fun onViewDetachedFromWindow(holder: RecyclerView.ViewHolder) {
        super.onViewDetachedFromWindow(holder)
        val viewHolder: RecyclerViewHolder<T> = holder as RecyclerViewHolder<T>
        viewHolder.widget.onDetachedFromWindow()
    }


    override fun onViewRecycled(holder: RecyclerView.ViewHolder) {
        super.onViewRecycled(holder)
        val viewHolder: RecyclerViewHolder<T> = holder as RecyclerViewHolder<T>
        viewHolder.widget.remove()
    }

    override fun onViewAttachedToWindow(holder: RecyclerView.ViewHolder) {
        super.onViewAttachedToWindow(holder)
        val viewHolder: RecyclerViewHolder<T> = holder as RecyclerViewHolder<T>
        viewHolder.widget.onAttachedToWindow()
    }

    protected fun bindData(widget: RecyclerViewHolderWidget<T>, position: Int) {
        widget.bindData(position, getItem(position))
    }

    fun onCreateViewHolderWidget(viewType: Int): RecyclerViewHolderWidget<T> {

        return recyclerWidget.onCreateViewHolderWidget(viewType)
    }





}